### PR TITLE
Fix expiring quotes

### LINKED ISSUE #1255: Quotes expire on gnosis network (swap via CoW)
**Description**
From time to time it happens that it is impossible to do a swap on Gnosis Chain via CoW.

#### Definition of done
- [ ] Description is created (Task creator)
- [ ] Code is complete (Developer)
- [ ] Scope of test and what was done section is defined in task (Developer)
- [ ] Review passed (Code reviewer)
- [ ] Verified manually (QA)
- [ ] Test cases created/updated if necessary (QA)
- [ ] Change successfully merged
- [ ] Release notes added if needed

**How to Reproduce**
1. Go to 'Swap' page
2. Select token from XDAI to USDC (for example)
3. Select 'Cow' exchange option 
4. Click on 'Wrap' button
5. Confirm wrap on Metamask
6. Click on 'Swap with CoW'
7. Click on 'Confirm Swap'
8. Click on 'Sing' (and you will get error message)

**Screenshots**
image

**Additional context**
This has not happened before and this was not the case earlier during testing.
Bug found on swap.site 1.0.0-beta.14.2

### PR DESCRIPTION
**Pull request by jorgedclopes**
# Summary

Fixes #1255

Add feature to recompute the trades periodically.
Set frequency to 5 minutes.

  # To Test

Go to 'Swap' page
Select token from XDAI to USDC (for example)
Select 'Cow' exchange option
Click on 'Wrap' button
Confirm wrap on Metamask
Click on 'Swap with CoW'
Click on 'Confirm Swap'
Click on 'Sign' (problem: and you will get error message; expected: no error message/successful transaction)

  # Background

COW/Gnosis trades quotes don't come with a TTL. In order to work around this, we will set it on our end.
It doesn't seem to be possible to compute the quote just for the COW trade and cascade that information without computing for all trades. Given that, we periodically compute for all trades.

### CODE REVIEW

**Comment by jorgedclopes** [comment]
Not fixed yet. Back to development.

**Comment by Diogomartf** [comment]
Still got the error in the preview!

image

File: src/state/swap/hooks.ts [hunk 1]
```diff
@@ -145,6 +145,19 @@ export function useDerivedSwapInfo(platformOverride?: RoutablePlatform): UseDeri
...
+  const currentTime = new Date().getTime()
+  // set TTL - currently at 2 minutes
+  const quoteTTL = 2 * 60 * 1000
+
+  const [endOfTime, setEndOfTime] = useState(currentTime + quoteTTL)
+  const [computeSwitch, setComputeSwitch] = useState(false)
+
+  if (currentTime > endOfTime) {
+    setComputeSwitch(!computeSwitch)
+    setEndOfTime(currentTime + quoteTTL)
```
Comment (adamazad) [inline src/state/swap/hooks.ts:158]: Setting state here causes re-render(s) of this function.

File: src/state/swap/hooks.ts [hunk 2]
```diff
@@ -145,6 +145,19 @@ export function useDerivedSwapInfo(platformOverride?: RoutablePlatform): UseDeri
   const parsedAmountString = `${parsedAmount?.currency.address?.toString()}-${parsedAmount?.raw?.toString()}`
   const recipientLookupComputed = `${recipientLookup.loading}-${recipientLookup?.address}-${recipientLookup?.name}`
 
+  // get UTC time in miliseconds
+  const currentTime = new Date().getTime()
+  // set TTL - currently at 2 minutes
+  const quoteTTL = 2 * 60 * 1000
+
+  const [endOfTime, setEndOfTime] = useState(currentTime + quoteTTL)
+  const [computeSwitch, setComputeSwitch] = useState(false)
```
Comment (adamazad) [inline src/state/swap/hooks.ts:154]: Use a timeOut ref that is set in `.then` of `getExactIn/Out`

**Review by adamazad — CHANGES_REQUESTED** [review]

**Review by adamazad — APPROVED** [review]
Works as expected. Just might make sense to think about the UX if anyone @zamli has any suggestions.

**Comment by niemam29** [comment]
For me this problem still sometimes occur, esp. when i am waiting a bit before swapping.
cach2

**Review by MilanVojnovic95 — APPROVED** [review]
Tested several times and everything seems ok. 👍

**Review by berteotti — APPROVED** [review]
LGTM, simple and clear

**Comment by adamazad** [comment]
This should be merged. The only thing I would like to see is that refreshing the quote should happen silently. The loading animation seems too distracting, at least. @zamli what do you think UX/UI wise?

**Pull request merged**